apps/vapi/RenderManager.cpp

```diff
@@ -0,0 +1,239 @@
+        float wa = wWidth / (float)wHeight;
+
+        if (fa >= wa) {
+            int x = 0;
+            int y = (wHeight / 2) - (wHeight / fa * wa / 2);
+            int w = wWidth;
+            int h = wHeight / fa * wa;
+            glViewport(x, y, w, h);
+        } else {
+            int x = (wWidth / 2) - (wWidth * fa / wa / 2);
+            int y = 0;
+            int w = wWidth * fa / wa;
+            int h = wHeight;
+            glViewport(x, y, w, h);
+        }
+    } else {
+        glViewport(0, 0, width, height);
+    }
+
+    mm->MatrixModeProjection();
+    mm->LoadIdentity();
+
+    float w = (float)width / (float)height;
+
+    if (vParams->GetProjectionType() == ViewpointParams::MapOrthographic) {
+        Trackball trackball;
+        double    center[3];
+        vParams->GetRotationCenter(center);
+        trackball.setFromFrame(posvec, dirvec, upvec, center, true);
+        float s = trackball.GetOrthoSize();
+        mm->Ortho(-s * w, s * w, -s, s, nearDist, farDist);
+        vParams->SetOrthoProjectionSize(trackball.GetOrthoSize());
+    } else {
+        double fov = vParams->GetFOV();
+        mm->Perspective(glm::radians(fov), w, nearDist, farDist);
+    }
+
+    double pMatrix[16];
+    mm->GetDoublev(MatrixManager::Mode::Projection, pMatrix);
+
+    bool enabled = _controlExec->GetSaveStateEnabled();
+    _controlExec->SetSaveStateEnabled(false);
+
+    vParams->SetProjectionMatrix(pMatrix);
+
+    _controlExec->SetSaveStateEnabled(enabled);
+
+    mm->MatrixModeModelView();
+}
+
+void RenderManager::setUpModelViewMatrix()
+{
+    ParamsMgr *      paramsMgr = _controlExec->GetParamsMgr();
+    ViewpointParams *vParams = paramsMgr->GetViewpointParams(getWinName());
+
+    double m[16];
+    vParams->GetModelViewMatrix(m);
+    _glManager->matrixManager->LoadMatrixd(m);
+}
+
+
+
+int RenderManager::Render(String imagePath)
+{
+    //    GL_ERR_BREAK();
+    if (!_glManager) _glManager = new GLManager;
+
+    static bool temp = false;
+    if (!temp) {
+        _controlExec->InitializeViz(getWinName(), _glManager);
+        temp = true;
+        // TODO: REMOVE TEMP
+    }
+
+    Framebuffer defaultFB;
+    defaultFB.Generate();
+    defaultFB.SetSize(600, 480);
+    defaultFB.MakeRenderTarget();
+
+    getViewpointParams()->SetWindowSize(600, 480);
+    //    getViewpointParams()->SetValueLong(ViewpointParams::UseCustomFramebufferTag, "", true);
+    //    getViewpointParams()->SetValueLong(ViewpointParams::CustomFramebufferWidthTag, "", 600);
+    //    getViewpointParams()->SetValueLong(ViewpointParams::CustomFramebufferHeightTag, "", 480);
+
+    _glManager->matrixManager->MatrixModeProjection();
+    _glManager->matrixManager->PushMatrix();
+    setUpProjMatrix();
+
+    _glManager->matrixManager->MatrixModeModelView();
+    _glManager->matrixManager->PushMatrix();
+    setUpModelViewMatrix();
+
+    int rc = _controlExec->EnableImageCapture(imagePath, getWinName());
```

Comment:
We're not handling error messages generated from MyBase::SetErrMsg() here.

Reply:
They are configured to automatically output to STDERR.